core/debug/src/debug_beetle.c

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifdef CORE_BEETLE
+
+#include <uvisor.h>
+
+#define UART1       0x40005000
+#define UART1DATA   (UART1)
+#define UART1STATE  (UART1 + 0x4)
+#define UART1CTRL   (UART1 + 0x8)
+#define UART1BDIV   (UART1 + 0x10)
+
+static inline uint32_t beetle_uart1_readl(uint32_t reg)
+{
+    return *(volatile uint32_t *)reg;
+}
+
+static inline void beetle_uart1_writel(uint32_t reg, uint32_t val)
+{
+    *(uint32_t *)reg = val;
```

Comment:
@fvincenzo form the looks of it you would be better off simply declaring a struct and casting it to the absolute offset. Each struct member would be declared with _IO _R _W as seen in the CMSIS headers - which would solve the "volatile"- problems.